internal/builtin/providers/terraform/state_store_inmem.go

```diff
@@ -0,0 +1,229 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package terraform
+
+import (
+	"errors"
+	"fmt"
+	"sort"
+	"sync"
+
+	"github.com/hashicorp/go-uuid"
+	"github.com/hashicorp/terraform/internal/backend"
+	"github.com/hashicorp/terraform/internal/configs/configschema"
+	"github.com/hashicorp/terraform/internal/providers"
+	"github.com/hashicorp/terraform/internal/tfdiags"
+	"github.com/zclconf/go-cty/cty"
+)
+
+// Matches default in command package.
+const defaultStateStoreChunkSize int64 = 8 << 20 // 8 MB
+
+// InMemStoreSingle allows 'storing' state in memory for the purpose of testing.
+//
+// "Single" reflects the fact that this implementation does not use any global scope vars
+// in its implementation, unlike the current inmem backend. HOWEVER, you can test whether locking
+// blocks multiple clients trying to access the same state at once by creating multiple instances
+// of backend.Backend that wrap the same provider.Interface instance.
+type InMemStoreSingle struct {
+	states stateMap
+	locks  lockMap
```

Comment:
For context, the `inmem` backend uses these global vars: 

https://github.com/hashicorp/terraform/blob/efb2a1c5676c71409265b46f978acf62e9d1437c/internal/backend/remote-state/inmem/backend.go#L24-L30

Like the code comment says, multiple backends can attempt to access the same state by using the same providers.Interface instance to create them.